services/src/main/scala/cromwell/pipeline/service/ProjectConfigurationService.scala

```diff
@@ -1,17 +1,33 @@
 package cromwell.pipeline.service
 
 import cromwell.pipeline.datastorage.dao.repository.DocumentRepository
-import cromwell.pipeline.datastorage.dto.{ ProjectConfiguration, ProjectId }
+import cromwell.pipeline.datastorage.dto.{ ProjectConfiguration, ProjectFileConfiguration, ProjectId }
 
 import scala.concurrent.{ ExecutionContext, Future }
 
 class ProjectConfigurationService(repository: DocumentRepository)(implicit ec: ExecutionContext) {
   import ProjectConfiguration._
 
-  def addConfiguration(projectConfiguration: ProjectConfiguration): Future[String] =
-    repository
-      .updateOne(toDocument(projectConfiguration), "projectId", projectConfiguration.projectId.value)
-      .map(_.toString)
+  def addConfiguration(projectConfiguration: ProjectConfiguration): Future[String] = {
+    def updateConfiguration(configuration: ProjectConfiguration) =
+      repository.updateOne(toDocument(configuration), "projectId", projectConfiguration.projectId.value).map(_.toString)
+
+    getById(projectConfiguration.projectId).flatMap {
+      case Some(config) if projectConfiguration.projectFileConfigurations != config.projectFileConfigurations =>
+        val oldFileConfigsMap =
+          config.projectFileConfigurations.flatMap(ProjectFileConfiguration.toMap).toMap
+        val newFileConfigsMap =
+          projectConfiguration.projectFileConfigurations.flatMap(ProjectFileConfiguration.toMap).toMap
+
+        val updatedFileConfigsMap = oldFileConfigsMap ++ newFileConfigsMap
+        val updatedFileConfigsSet =
+          updatedFileConfigsMap.flatMap(x => Set(ProjectFileConfiguration.fromMap(x._1, x._2))).toSet
```

Comment:
Avoid meaningless variables especially when you work with tuples. It's hard to understand what `x._2` means.
```suggestion
          updatedFileConfigsMap.flatMap { case (path, inputs) => Set(ProjectFileConfiguration.fromMap(path, inputs))).toSet
```

Reply:
Very useful, thanks!
Fixed.